### PR TITLE
backwards for softcapping

### PR DESCRIPTION
**Pull request by lucidrains**
backwards pass for softcapping

dtanh is stored right after softcapping is applied, which is then later multiplied to `dS`

### CODE REVIEW

**Comment by lucidrains** [comment]
@tridao whatever you think is the best way let's roll with that!

File: csrc/flash_attn/src/flash_bwd_kernel.h [hunk 1]
```diff
@@ -567,16 +591,34 @@ inline __device__ void compute_dq_dk_dv_1colblock(const Params &params, const in
...
+            auto pointwise_mult = [](float p, float dp, float d, float m) {
+                return p * (!Is_dropout || p >= 0 ? dp - d : d) * m;
+            };
+            #pragma unroll
+            for (int mi = 0; mi < size<0>(dS); ++mi) {
+                const float max_scaled = lse(mi) == -INFINITY ? 0.f : lse(mi) * float(M_LOG2E);
+                #pragma unroll
+                for (int ni = 0; ni < size<1>(dS); ++ni) {
+                    const float post_softcap_value = (log2f(max(scores(mi, ni), 1e-20)) + max_scaled) / params.scale_softmax_log2;
+                    const float dtanh = (1. - post_softcap_value * post_softcap_value);
```
Comment (lucidrains) [inline csrc/flash_attn/src/flash_bwd_kernel.h]: going from scores back to output of `apply_softcap` to calculate dtanh

File: csrc/flash_attn/src/flash_bwd_kernel.h [hunk 2]
```diff
@@ -471,10 +483,22 @@ inline __device__ void compute_dq_dk_dv_1colblock(const Params &params, const in
...
+
         // Reshape acc_s from (MMA=4, MMA_N, MMA_N) to (row=(2, MMA_N), col=(2, MMA_N))
         Tensor scores = make_tensor(acc_s.data(), flash::convert_layout_acc_rowcol(acc_s.layout()));
         // if (cute::thread(32, 0)) { print(scores); }
 
+        const bool store_dtanh = true;  // toggles between the two ways of approaching backwards
+        Tensor dtanh = make_tensor_like(scores);
+
+        if (Is_softcap && store_dtanh) {
+            cute::copy(scores, dtanh);
```
Comment (lucidrains) [inline csrc/flash_attn/src/flash_bwd_kernel.h]: copies scores and calculates dtanh for later scaling of `dS`

**Comment by tridao** [comment]
My guess is that it's better to store the dtanh. Calculating the inverse with log would be a lot slower.

File: csrc/flash_attn/src/flash_bwd_kernel.h [hunk 3]
```diff
@@ -567,16 +591,34 @@ inline __device__ void compute_dq_dk_dv_1colblock(const Params &params, const in
...
+            for (int mi = 0; mi < size<0>(dS); ++mi) {
+                const float max_scaled = lse(mi) == -INFINITY ? 0.f : lse(mi) * float(M_LOG2E);
+                #pragma unroll
+                for (int ni = 0; ni < size<1>(dS); ++ni) {
+                    const float post_softcap_value = (log2f(max(scores(mi, ni), 1e-20)) + max_scaled) / params.scale_softmax_log2;
+                    const float dtanh = (1. - post_softcap_value * post_softcap_value);
+                    dS(mi, ni) = pointwise_mult(scores(mi, ni), dS(mi, ni), dP_sum(mi), dtanh);
+                }
+            }
+        } else {
```
Comment (winglian) [inline csrc/flash_attn/src/flash_bwd_kernel.h]: Should this be else if is_softcap? Otherwise this would run even if softcap isn't used.

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: @winglian hey Wing! so that actually exists at line 618 `Is_softcap ? dtanh : 1.f`

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: it'll look a lot cleaner once we move to one way of doing it

**Comment by lucidrains** [comment]
> My guess is that it's better to store the dtanh. Calculating the inverse with log would be a lot slower.

sounds good :pray: 

i'll get this wrapped up today and have you and @Narsil do a review of my probably really bad cpp code :laughing:

**Comment by lucidrains** [comment]
@tridao @Narsil ok, i've removed all the local stuff to make compiling faster, currently doing the full half hour compiling to run the entire test suite

let me know if you see any big issues with the code, esp how the constexpr ternary was handled

File: csrc/flash_attn/src/flash_bwd_kernel.h [hunk 4]
```diff
@@ -567,14 +584,23 @@ inline __device__ void compute_dq_dk_dv_1colblock(const Params &params, const in
...
         for (int mi = 0; mi < size<0>(dS); ++mi) {
             #pragma unroll
             for (int ni = 0; ni < size<1>(dS); ++ni) {
-                dS(mi, ni) = pointwise_mult(scores(mi, ni), dS(mi, ni), dP_sum(mi));
+
+                float maybe_dtanh;
+                if constexpr (Is_softcap) {
+                    maybe_dtanh = dtanh(mi, ni);
+                } else {
+                    maybe_dtanh = 1.f;
```
Comment (lucidrains) [inline csrc/flash_attn/src/flash_bwd_kernel.h]: if the compiler isn't smart enough here, i guess can do it the verbose way?

```cpp
        if constexpr (Is_softcap) {
            auto pointwise_mult = [](float p, float dp, float d, float m) {
                return p * (!Is_dropout || p >= 0 ? dp - d : d) * m;
            };
            #pragma unroll
            for (int mi = 0; mi < size<0>(dS); ++mi) {
                #pragma unroll
                for (int ni = 0; ni < size<1>(dS); ++ni) {
                    dS(mi, ni) = pointwise_mult(scores(mi, ni), dS(mi, ni), dP_sum(mi), dtanh(mi, ni));
                }
            }
        } else {
            auto pointwise_mult = [](float p, float dp, float d) {
                return p * (!Is_dropout || p >= 0 ? dp - d : d);
            };
            #pragma unroll
            for (int mi = 0; mi < size<0>(dS); ++mi) {
                #pragma unroll
                for (int ni = 0; ni < size<1>(dS); ++ni) {
                    dS(mi, ni) = pointwise_mult(scores(mi, ni), dS(mi, ni), dP_sum(mi));
                }
            }
        }
```

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: oh actually, another way would be

```cpp
        auto pointwise_mult = [](float p, float dp, float d) {
            return p * (!Is_dropout || p >= 0 ? dp - d : d);
        };
        #pragma unroll
        for (int mi = 0; mi < size<0>(dS); ++mi) {
            #pragma unroll
            for (int ni = 0; ni < size<1>(dS); ++ni) {

                float scaled_ds = pointwise_mult(scores(mi, ni), dS(mi, ni), dP_sum(mi));

                if constexpr (Is_softcap) {
                    scaled_ds = scaled_ds * dtanh(mi, ni);
                }

                dS(mi, ni) = scaled_ds;
            }
        }
```

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: @Narsil i think you may know the answer to which way is best here

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: actually curious if a large language / code model can answer which way is best if i present all the options lol

Comment (lucidrains) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: i'll go with the last option

Comment (tridao) [reply csrc/flash_attn/src/flash_bwd_kernel.h]: that looks good to me

**Comment by lucidrains** [comment]
been waiting for over an hour for the full settings to compile now :cry:

**Comment by tridao** [comment]
On compile time: yea it's getting pretty bad because of all the templating. Takes about 12m on a 176-core machine. I might start dropping some features.

**Comment by lucidrains** [comment]
@tridao 
compiling

**Comment by lucidrains** [comment]
left a machine on overnight to compile, and when i woke up, it was still compiling

not sure what's going on, but i'm just going to test head dimension of 64 and call it a day

**Comment by tridao** [comment]
I think when there are more than some number templates, it actually takes way way longer.
e.g. I had to split to more .cu files to compile the softcap fwd successfully: https://github.com/Dao-AILab/flash-attention/commit/908511b2b6144d394795fa82d5179ce04a9479d8

This is clearly not sustainable, we'll fix it in the next couple of weeks.

**Comment by lucidrains** [comment]
> I think when there are more than some number templates, it actually takes way way longer. e.g. I had to split to more .cu files to compile the softcap fwd successfully: [908511b](https://github.com/Dao-AILab/flash-attention/commit/908511b2b6144d394795fa82d5179ce04a9479d8)
> 
> This is clearly not sustainable, we'll fix it in the next couple of weeks.

yea no problem! i just spun up a new box and did the whole head dimension filtering + disable flags and got it down to 5 minutes again

found a bug 🤦 but otherwise should pass now if you want to give it a more extensive test run!

**Comment by lucidrains** [comment]
@tridao after we land this, i can also spend some time and introduce a pull request to automate the head dimension filtering and disable flag

the way i imagine it is some config file at the root directory

```json
{
  "head_dimensions": [64, 128],
  "enable_only": ["causal", "softcapping"]
}
```

and one can just do `make build_and_run_test` and it will only compile the relevant `.cu` as well as the appropriate forwards and backwards. the hardest part may be this `static_switch.h` file, but with some jinja templating, i think it is doable.

**Comment by lucidrains** [comment]
@tridao flash attention 3?! 🤯

**Comment by Shreya-Pathak** [comment]
Hi, do we have an estimate on when this change will be integrated? And what are the current blockers?

**Comment by lucidrains** [comment]
@Shreya-Pathak works for me, just waiting on a review and merge

**Comment by Shreya-Pathak** [comment]
I'm from the Gemma team and we wanted to have support for softcapping in Flash Attention as soon as possible as it is affecting a lot of downstream users. Happy to contribute if needed.

**Comment by lucidrains** [comment]
@Shreya-Pathak :exploding_head: surprise reveal lol

great job with Gemma! thank you for open sourcing :smile:

**Comment by Lzhang-hub** [comment]
build form source cost so much time. how to speedup it?

**Comment by tridao** [comment]
I'll try to review and merge tmr

**Comment by zhengyaowei** [comment]
It works for me when fine-tuning Gemma2 with flash-attn, while the `flash_attn==2.6.1` produces NaN grads in backwards.

**Comment by kostum123** [comment]
Could we please merge this soon? We have multiple confirmations that it is working correctly.

**Comment by seungduk-yanolja** [comment]
https://github.com/Dao-AILab/flash-attention/blob/e4769d7a1ad5171dce7a3fb3cbddd6204e113b4d/flash_attn/flash_attn_interface.py#L78
There is no softcap argument but

https://github.com/Dao-AILab/flash-attention/blob/e4769d7a1ad5171dce7a3fb3cbddd6204e113b4d/flash_attn/flash_attn_interface.py#L621
It calls with softcap

**Comment by seungduk-yanolja** [comment]
The issue in `flash_attn_interface.py` is resolved when sync'ed to the original main branch.
However, facing a new error like this:
```
[rank6]:   File "/home/seungduk/miniconda3/envs/axo/lib/python3.10/site-packages/transformers/models/gemma2/modeling_gemma2.py", line 361, in forward
[rank6]:     attn_output = _flash_attention_forward(
[rank6]:   File "/home/seungduk/miniconda3/envs/axo/lib/python3.10/site-packages/transformers/modeling_flash_attention_utils.py", line 212, in _flash_attention_forward
[rank6]:     attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
[rank6]:   File "/home/seungduk/miniconda3/envs/axo/lib/python3.10/site-packages/flash_attn-2.6.1-py3.10-linux-x86_64.egg/flash_attn/bert_padding.py", line 212, in pad_input
[rank6]:     output = index_put_first_axis(hidden_states, indices, batch * seqlen)
[rank6]:   File "/home/seungduk/miniconda3/envs/axo/lib/python3.10/site-packages/torch/autograd/function.py", line 598, in apply
[rank6]:     return super().apply(*args, **kwargs)  # type: ignore[misc]
[rank6]:   File "/home/seungduk/miniconda3/envs/axo/lib/python3.10/site-packages/flash_attn-2.6.1-py3.10-linux-x86_64.egg/flash_attn/bert_padding.py", line 51, in forward
[rank6]:     output[indices] = values
```

Setup: H100x8x2 and Axolotl, transformers was installed from the latest source

**Comment by winglian** [comment]
@tridao will we get a 2.6.2 release w/ wheels for this?

**Comment by tridao** [comment]
Ya, today or tmr. I'm trying to figure out the compile time issue so the github runner doesn't time out.

**Comment by lucidrains** [comment]
thank you Tri! understand you must be swamped

I'll send a PR later this week for that development config idea above

**Pull request merged**